packages/react-components/src/components/TextFieldWithMention/TextFieldWithMention.tsx

```diff
@@ -0,0 +1,629 @@
+            activeSuggestionIndex === undefined
+              ? mentionSuggestions.length - 1
+              : Math.max(activeSuggestionIndex - 1, 0);
+          setActiveSuggestionIndex(newActiveIndex);
+        }
+      } else if (ev.key === 'ArrowDown') {
+        ev.preventDefault();
+        if (mentionSuggestions.length > 0) {
+          const newActiveIndex =
+            activeSuggestionIndex === undefined
+              ? 0
+              : Math.min(activeSuggestionIndex + 1, mentionSuggestions.length - 1);
+          setActiveSuggestionIndex(newActiveIndex);
+        }
+      }
+      if (ev.key === 'Enter' && (ev.shiftKey === false || !supportNewline)) {
+        ev.preventDefault();
+        // If we are looking up a mention, select the focused suggestion
+        if (mentionSuggestions.length > 0 && activeSuggestionIndex !== undefined) {
+          const selectedMention = mentionSuggestions[activeSuggestionIndex];
+          if (selectedMention) {
+            onSuggestionSelected(selectedMention);
+            return;
+          }
+        }
+        onEnterKeyDown && onEnterKeyDown();
+      }
+      onKeyDown && onKeyDown(ev);
+    },
+    [onEnterKeyDown, onKeyDown, supportNewline, mentionSuggestions, activeSuggestionIndex, onSuggestionSelected]
+  );
+
+  const debouncedQueryUpdate = useDebouncedCallback(async (query?: string) => {
+    if (query === undefined) {
+      updateMentionSuggestions([]);
+      return;
+    }
+    const suggestions = (await mentionLookupOptions?.onQueryUpdated(query)) ?? [];
+    if (suggestions.length === 0) {
+      setActiveSuggestionIndex(undefined);
+    } else if (activeSuggestionIndex === undefined) {
+      setActiveSuggestionIndex(0);
+    }
+    updateMentionSuggestions(suggestions);
+  }, 500);
+
+  useEffect(() => {
+    return () => {
+      debouncedQueryUpdate.cancel();
+    };
+  }, [debouncedQueryUpdate]);
+
+  const getUpdatedSelectionStartIndexOrEndIndex = useCallback(
+    (
+      inputTextValue: string,
+      tagsValue: TagData[],
+      selectionIndex: number,
+      selectionValue?: number
+    ): number | undefined => {
+      const mentionTag = findMentionTagForSelection(tagsValue, selectionIndex);
+      // don't include boundary cases to show correct selection, otherwise it will show selection at mention boundaries
+      if (
+        mentionTag !== undefined &&
+        mentionTag.plainTextBeginIndex !== undefined &&
+        selectionIndex > mentionTag.plainTextBeginIndex &&
+        selectionIndex < (mentionTag.plainTextEndIndex ?? mentionTag.plainTextBeginIndex)
+      ) {
+        // get updated selection index
+        const newSelectionIndex = findNewSelectionIndexForMention({
+          tag: mentionTag,
+          textValue: inputTextValue,
+          currentSelectionIndex: selectionIndex,
+          previousSelectionIndex: selectionValue ?? inputTextValue.length
+        });
+        return newSelectionIndex;
+      }
+      return undefined;
+    },
+    []
```

Comment:
Looks like no dependencies are needed for this function, consider move it outside of the component function instead of using useCallback

To make give them a rank:

const function outside of component body (if possible) > useCallback > functions without useCallback